types/react-dom/experimental.d.ts

```diff
@@ -50,4 +50,23 @@ declare module '.' {
     type FormStatus = FormStatusPending | FormStatusNotPending;
 
     function experimental_useFormStatus(): FormStatus;
+
+    function experimental_useFormState<State>(
+        action: (state: State, formData: FormData) => Promise<State>,
+        initialState: State,
+        permalink?: string,
+    ): [state: State, dispatch: (formData: FormData) => void];
+}
+
+declare module './client' {
+    type ReactFormState<State, ReferenceId> = [
```

Comment:
This part is an opaque type so shouldn’t really be defined here and its properties would likely constantly be out of date.

It also doesn’t need the parameterization. I think this is actually not the correct parameterization in React internally neither because the ReferenceId that can pass here comes from more than one build. The state could also change to be multiple which is opaque too.

Could just make it some nonsense type to make it nominal. Eg `FormState = { __OPAQUE_FORM_STATE: true }`.

Reply:
Without something producing this type, it won't be useful at the moment anyway. But at least you now know if you typed the option correct.